server/service/devices.go

```diff
@@ -335,3 +337,79 @@ func transparencyURL(ctx context.Context, request interface{}, svc fleet.Service
 
 	return transparencyURLResponse{RedirectURL: transparencyURL}, nil
 }
+
+////////////////////////////////////////////////////////////////////////////////
+// Get Current Device's MDM Apple Enrollment Profile
+////////////////////////////////////////////////////////////////////////////////
+
+type getDeviceMDMManualEnrollProfileRequest struct {
+	Token string `url:"token"`
+}
+
+func (r *getDeviceMDMManualEnrollProfileRequest) deviceAuthToken() string {
+	return r.Token
+}
+
+type getDeviceMDMManualEnrollProfileResponse struct {
+	// Profile field is used in hijackRender for the response.
+	Profile []byte
+
+	Err error `json:"error,omitempty"`
+}
+
+func (r getDeviceMDMManualEnrollProfileResponse) hijackRender(ctx context.Context, w http.ResponseWriter) {
+	// TODO(mna): does this filename look ok?
+	// make the browser download the content to a file
+	w.Header().Add("Content-Disposition", `attachment; filename="fleet-mdm-enrollment-profile.mobileconfig"`)
+	// explicitly set the content length before the write, so the caller can
+	// detect short writes (if it fails to send the full content properly)
+	w.Header().Set("Content-Length", strconv.FormatInt(int64(len(r.Profile)), 10))
+	// this content type will make macos open the profile with the proper application
+	w.Header().Set("Content-Type", "application/x-apple-aspen-config; charset=urf-8")
+	// prevent detection of content, obey the provided content-type
+	w.Header().Set("X-Content-Type-Options", "nosniff")
+
+	if n, err := w.Write(r.Profile); err != nil {
+		logging.WithExtras(ctx, "err", err, "written", n)
+	}
+}
+
+func (r getDeviceMDMManualEnrollProfileResponse) error() error { return r.Err }
+
+func getDeviceMDMManualEnrollProfileEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
+	// this call ensures that the authentication was done, no need to actually
+	// use the host
+	if _, ok := hostctx.FromContext(ctx); !ok {
+		err := ctxerr.Wrap(ctx, fleet.NewAuthRequiredError("internal error: missing host from request context"))
+		return getDeviceMDMManualEnrollProfileResponse{Err: err}, nil
+	}
+
+	profile, err := svc.GetDeviceMDMAppleEnrollmentProfile(ctx)
+	if err != nil {
+		return getDeviceMDMManualEnrollProfileResponse{Err: err}, nil
+	}
+	return getDeviceMDMManualEnrollProfileResponse{Profile: profile}, nil
+}
+
+func (svc *Service) GetDeviceMDMAppleEnrollmentProfile(ctx context.Context) ([]byte, error) {
+	// skipauth: service is called by device-authenticated endpoint, adding
+	// SkipAuth call here just to indicate explicitly that no additional
+	// authorization is required.
+	svc.authz.SkipAuthorization(ctx)
```

Comment:
thinking out loud: should we signal this by using `svc.authz.IsAuthenticatedWith(ctx, authz.AuthnDeviceToken)` instead?

Reply:
Yeah that's a good idea.